causallearn/utils/cit.py

```diff
@@ -15,78 +15,142 @@
+            with codecs.open(self.cache_path, 'w') as fout: fout.write(json.dumps(self.pvalue_cache, indent=2))
+            self.last_time_cache_saved = time.time()
+
+    def get_formatted_XYZ_and_cachekey(self, X, Y, condition_set):
+        '''
+        reformat the input X, Y and condition_set to
+            1. convert to built-in types for json serialization
+            2. handle multi-dim unconditional variables (for kernel-based)
+            3. basic check for valid input (X, Y no overlap with condition_set)
+            4. generate unique and hashable cache key
+
+        Parameters
+        ----------
+        X: int, or np.*int*
+        Y: int, or np.*int*
+        condition_set: Iterable<int | np.*int*>
+
+        Returns
+        -------
+        Xs: List<int>, sorted. may swapped with Ys for cache key uniqueness.
+        Ys: List<int>, sorted.
+        condition_set: List<int>
+        cache_key: string. Unique for <X,Y|S> in any input type or order.
+        '''
+        def _stringize(ulist1, ulist2, clist):
+            # ulist1, ulist2, clist: list of ints, sorted.
+            _strlst  = lambda lst: '.'.join(map(str, lst))
+            return f'{_strlst(ulist1)};{_strlst(ulist2)}|{_strlst(clist)}' if len(clist) > 0 else \
+                   f'{_strlst(ulist1)};{_strlst(ulist2)}'
+
+        # every time when cit is called, auto save to local cache.
+        self.save_to_local_cache()
+
+        METHODS_SUPPORTING_MULTIDIM_DATA = ["kci"]
+        if condition_set is None: condition_set = []
+        # 'int' to convert np.*int* to built-in int; 'set' to remove duplicates; sorted for hashing
+        condition_set = sorted(set(map(int, condition_set)))
+
+        # usually, X and Y are 1-dimensional index (in constraint-based methods)
+        if self.method not in METHODS_SUPPORTING_MULTIDIM_DATA:
+            X, Y = (int(X), int(Y)) if (X < Y) else (int(Y), int(X))
+            assert X not in condition_set and Y not in condition_set, "X, Y cannot be in condition_set."
+            return [X], [Y], condition_set, _stringize([X], [Y], condition_set)
+
+        # also to support multi-dimensional unconditional X, Y (usually in kernel-based tests)
+        Xs = sorted(set(map(int, X))) if isinstance(X, Iterable) else [int(X)]  # sorted for comparison
+        Ys = sorted(set(map(int, Y))) if isinstance(Y, Iterable) else [int(Y)]
+        Xs, Ys = (Xs, Ys) if (Xs < Ys) else (Ys, Xs)
+        assert len(set(Xs).intersection(condition_set)) == 0 and \
+               len(set(Ys).intersection(condition_set)) == 0, "X, Y cannot be in condition_set."
+        return Xs, Ys, condition_set, _stringize(Xs, Ys, condition_set)
+
+class FisherZ(CIT_Base):
+    def __init__(self, data, **kwargs):
+        super().__init__(data, **kwargs)
+        self.check_cache_method_consistent('fisherz', -1)   # -1: no parameters can be specified for fisherz
```

Comment:
hmm, -1 here looks ugly. maybe we can have better design, e.g. use None?

And in the code below, it seems you never write "parameters_hash" to json? Why the assertion not fail?

Reply:
Yeh -1 is ugly, lol. A message string ("NO SPECIFIED PARAMETERS") might also be ok.

"parameters_hash" is written to cache (and json). See [`check_cache_method_consistent`](https://github.com/cmu-phil/causal-learn/blob/3781daec48142c53b0243dc000d015fcc2498fb3/causallearn/utils/cit.py#L65).

Reply:
yeah, a (const) string is much better.

try to avoid to use any magic number like -1 here in your code. :)